monitoring/ping.go

```diff
@@ -0,0 +1,308 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/satellite/agent/health"
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+
+	"github.com/codahale/hdrhistogram"
+	"github.com/gravitational/trace"
+	"github.com/gravitational/ttlmap"
+	serf "github.com/hashicorp/serf/client"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// pingCheckerID specifies the check name
+	pingCheckerID = "ping-checker"
+	// slidingWindowSize specifies the number of retained check results
+	slidingWindowSize = 10
+	// statsTTLPeriod specifies how long check results will be kept before being dropped
+	statsTTLPeriod = 1 * time.Hour
+	// pingMinimum sets the minimum value that can be recorded
+	pingMinimum = 0 * time.Second
+	// pingMaximum sets the maximum value that can be recorded
+	pingMaximum = 10 * time.Second
+	// pingSignificantFigures specifies how many decimals should be recorded
+	pingSignificantFigures = 3
+	// latencyThreshold sets the RTT threshold
+	latencyThreshold = 15 * time.Millisecond
+	// latencyQuantile sets the quantile used while checking Histograms against RTT results
+	latencyQuantile = 95.0
+)
+
+// NewPingChecker returns a checker that verifies accessibility of nodes in the cluster by exchanging ping requests
+func NewPingChecker(serfRPCAddr string, serfMemberName string) (c health.Checker, err error) {
+	latencyTTLMap, err := ttlmap.New(int(statsTTLPeriod.Seconds())) //FIXME: why is number of seconds used as capacity?
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	log.WithFields(log.Fields{trace.Component: "ping"})
+	log.Debugf("using Serf IP: %v", serfRPCAddr)
+	log.Debugf("using Serf Name: %v", serfMemberName)
+	// fetch serf config and instantiate client
+	clientConfig := serf.Config{
```

Comment:
I don't think we should implement ALL features now, otherwise this task will never end and will be absorbed in a never ending whirlwind of scope-creep.

I can definitely add it as a TODO comment, along with the other desired features but I don't think it should be implemented now if we don't have a clear, pragmatic, real application that we use right now.